src/hotspot/share/code/nmethod.cpp

```diff
@@ -474,13 +490,30 @@ nmethod* nmethod::new_native_nmethod(const methodHandle& method,
     if (exception_handler != -1) {
       offsets.set_value(CodeOffsets::Exceptions, exception_handler);
     }
-    nm = new (native_nmethod_size, CompLevel_none)
+
+    // MH intrinsics are dispatch stubs which are compatible with NonNMethod space.
+    bool allow_NonNMethod_space = method->is_method_handle_intrinsic();
+    nm = new (native_nmethod_size, allow_NonNMethod_space)
     nmethod(method(), compiler_none, native_nmethod_size,
             compile_id, &offsets,
             code_buffer, frame_size,
             basic_lock_owner_sp_offset,
             basic_lock_sp_offset,
             oop_maps);
+#ifdef ASSERT
```

Comment:
How about making this whole block a separate function?

Reply:
Done. And thanks for the hint regarding `is_permanent_class_loader_data`. That was what I was missing. Assertion added.